algebra/lra.elpi

```diff
@@ -0,0 +1,1338 @@
+ground-hexadecimal {{ Hexadecimal.D0 lp:D }} :- !, ground-hexadecimal D.
+ground-hexadecimal {{ Hexadecimal.D1 lp:D }} :- !, ground-hexadecimal D.
+ground-hexadecimal {{ Hexadecimal.D2 lp:D }} :- !, ground-hexadecimal D.
+ground-hexadecimal {{ Hexadecimal.D3 lp:D }} :- !, ground-hexadecimal D.
+ground-hexadecimal {{ Hexadecimal.D4 lp:D }} :- !, ground-hexadecimal D.
+ground-hexadecimal {{ Hexadecimal.D5 lp:D }} :- !, ground-hexadecimal D.
+ground-hexadecimal {{ Hexadecimal.D6 lp:D }} :- !, ground-hexadecimal D.
+ground-hexadecimal {{ Hexadecimal.D7 lp:D }} :- !, ground-hexadecimal D.
+ground-hexadecimal {{ Hexadecimal.D8 lp:D }} :- !, ground-hexadecimal D.
+ground-hexadecimal {{ Hexadecimal.D9 lp:D }} :- !, ground-hexadecimal D.
+ground-hexadecimal {{ Hexadecimal.Da lp:D }} :- !, ground-hexadecimal D.
+ground-hexadecimal {{ Hexadecimal.Db lp:D }} :- !, ground-hexadecimal D.
+ground-hexadecimal {{ Hexadecimal.Dc lp:D }} :- !, ground-hexadecimal D.
+ground-hexadecimal {{ Hexadecimal.Dd lp:D }} :- !, ground-hexadecimal D.
+ground-hexadecimal {{ Hexadecimal.De lp:D }} :- !, ground-hexadecimal D.
+ground-hexadecimal {{ Hexadecimal.Df lp:D }} :- !, ground-hexadecimal D.
+
+pred ground-uint i:term.
+ground-uint {{ Number.UIntDecimal lp:D }} :- !, ground-decimal D.
+ground-uint {{ Number.UIntHexadecimal lp:D }} :- !, ground-hexadecimal D.
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+% Registering MathComp instances and sort coercions
+
+pred register-instance i:scope, i:id, i:gref, i:gref, i:constant -> prop.
+register-instance Scope DbName Proj Pat Pred :- std.do! [
+  coq.CS.db-for Proj (cs-gref Pat) [cs-instance _ _ (const Inst)],
+  coq.elpi.accumulate Scope DbName (clause _ _ (Pred Inst :- !)) ].
+
+pred canonical-init i:scope, i:id.
+canonical-init Scope DbName :- std.do! [
+  register-instance Scope DbName
+    {{:gref GRing.Nmodule.sort }} {{:gref nat }} canonical-nat-nmodule,
+  register-instance Scope DbName
+    {{:gref GRing.PzSemiRing.sort }} {{:gref nat }} canonical-nat-semiring,
+  register-instance Scope DbName
+    {{:gref GRing.ComPzSemiRing.sort }} {{:gref nat }} canonical-nat-comsemiring,
+  register-instance Scope DbName
+    {{:gref GRing.Nmodule.sort }} {{:gref int }} canonical-int-nmodule,
+  register-instance Scope DbName
+    {{:gref GRing.Zmodule.sort }} {{:gref int }} canonical-int-zmodule,
+  register-instance Scope DbName
+    {{:gref GRing.PzSemiRing.sort }} {{:gref int }} canonical-int-semiring,
+  register-instance Scope DbName
+    {{:gref GRing.PzRing.sort }} {{:gref int }} canonical-int-ring,
+  register-instance Scope DbName
+    {{:gref GRing.ComPzRing.sort }} {{:gref int }} canonical-int-comring,
+  register-instance Scope DbName
+    {{:gref GRing.UnitRing.sort }} {{:gref int }} canonical-int-unitring ].
+
+pred coercion-init i:scope, i:id.
+coercion-init Scope DbName :- std.do! [
+  coq.typecheck {{ @Order.le }} (prod _ _ DOrd) ok,
```

Comment:
This  `coq.typecheck X Y ok` and all subsequent ones should ideally be replaced by `std.assert-ok! (coq.typecheck X Y) "custom error message for debugging purposes`.